docs/contributing/error-handling.md

```diff
@@ -0,0 +1,366 @@
+
+If an AWS Go SDK service API is missing an error code constant, an AWS Support case should be submitted and a new constant can be added to `aws/internal/service/{SERVICE}/errors.go` file (created if not present), e.g.
+
+```go
+const(
+    ErrCodeInvalidParameterException = "InvalidParameterException"
+)
+```
+
+Then referencing code can use it via:
+
+```go
+// imports
+tf{SERVICE} "github.com/terraform-providers/terraform-provider-aws/aws/internal/service/{SERVICE}"
+
+// logic
+tfawserr.ErrCodeEquals(err, tf{SERVICE}.ErrCodeInvalidParameterException)
+```
+
+e.g.
+
+```go
+// imports
+tfec2 "github.com/terraform-providers/terraform-provider-aws/aws/internal/service/ec2"
+
+// logic
+tfawserr.ErrCodeEquals(err, tfec2.ErrCodeInvalidParameterException)
+```
+
+### Terraform Plugin SDK Types and Helpers
+
+The Terraform Plugin SDK includes some error types which are used in certain operations and typically preferred over implementing new types:
+
+* [`resource.NotFoundError`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#NotFoundError)
+* [`resource.TimeoutError`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#TimeoutError): Returned from [`resource.Retry()`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#Retry), [`resource.RetryContext()`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#RetryContext), [`(resource.StateChangeConf).WaitForState()`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#StateChangeConf.WaitForState), and [`(resource.StateChangeConf).WaitForStateContext()`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#StateChangeConf.WaitForStateContext)
+
+The Terraform AWS Provider codebase implements some additional helpers for working with these in the `github.com/terraform-providers/terraform-provider-aws/aws/internal/tfresource` package:
+
+- `tfresource.NotFound(err)`: Returns true if the error is a `resource.NotFoundError`.
+- `tfresource.TimedOut(err)`: Returns true if the error is a `resource.TimeoutError` and contains no `LastError`. This typically signifies that the retry logic was never signaled for a retry, which can happen when AWS Go SDK operations are automatically retrying before returning.
+
+## Resource Lifecycle Guidelines
+
+Terraform CLI and the Terraform Plugin SDK have certain expectations and automatic behaviors depending on the lifecycle operation of a resource. This section highlights some common issues that can occur and their expected resolution.
+
+### Resource Creation
+
+Invoked in the resource via the `schema.Resource` type `Create`/`CreateContext` function.
```

Comment:
with reference to `Context` functions, this might introduce the need to mention returning the `diag.Errorf()/diag.FromErr()` .. but maybe as a future enhancement to the content here 👍

Reply:
Ah yeah, I was getting ahead of myself there. Let us discuss this topic the next time we talk about the provider context project. It might be best to be opinionated upfront with the newer style and ensure that we can catch things via static analysis. 👍